Added basic support for USART/UART peripherals on STM32F4.

dimdimdimdim opened the pull request:
So, i've managed to do something for the USART/UART which seems to work (at least for UART1, both TX and RX). This is only for F4 currently, but it really shouldn't be hard to adapt it to the other families (there seems to be no difference in the behavior of these peripherals across families).
I don't know if you will be interested by this pull request, as you are planning some other ways of generating peripherals, but anyway...
And, indeed, after seeing your code with the complete register description and bit fields (e.g. for the RCC), I understand you want to automate this. I, myself, haven't been brave enough to describe these in details, so it is just described as a flat register peripheral (no bit field description). It's better than nothing.
However, regarding the "switch case statement", I have seen 8/16/32 bit register access is already implemented in register-peripheral.c, and for the unaligned accesses, I thought it could be easily solved by calling twice the read/write callbacks, once for each register affected by the unaligned operation (adjusting the offset and size appropriately for each call).

Note: there are a few other minor changes I made, you can see the details in the commit description.

Dimitri

ilg-ul commented:
oops! I completely missed this :-(

is it still up to date?

dimdimdimdim commented:
Up to date ? Do you mean: is it based on the current dev branch ? It 
seems yes, as there hasn't been any commits on gnuarmeclipse since my fork.


Le 17/02/2016 09:45, Liviu Ionescu a écrit :
>
> oops! I completely missed this :-(
>
> is it still up to date?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/gnuarmeclipse/qemu/pull/6#issuecomment-185099581>.
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

ilg-ul commented:
> Up to date ? Do you mean: is it based on the current dev branch ?

no, does it reflect your latest version, or since November you further enhanced your code?

dimdimdimdim commented:
No, I did not work on that code since the pull request. I planned to add timers peripheral, but I've been busy with other things. It is up to date.

ilg-ul commented:
ok, I'll try to check your patch and integrate it, hopefully in a few days.

tgbloks commented:
This was indeed easy to port to the F1 family. thx ;-)
I could create another pull request after this was merged...

ilg-ul commented:
Dimitri,

I finally processed your code. Most of it is already in, although not tested and for sure requiring fine tuning.

I had to manually add the patches, since the pull request in it's entirety was not ok.

One major remark: I did not understand the reason why you added the post_read callback, normally the logic should be implemented in the pre_read callback and the value returned. Then the common code should store this value in the register, and the read should return it (masked and shifted) from there. Similarly, but mirrored, for write, the value is written by the user to the register storage and post_write is invoked.

Could you fix these details and post the changes in a separate pull request? 

If you find multiple unrelated issues, please post them as distinct pul requests, otherwise one wrong commit will prevent the acceptance of the entire pull request.

Thank you,

Liviu

ilg-ul commented:
I took a second look at your USART, it seems very specific to F4, please be aware that other families have different definitions, for example F1 is completely different.

dimdimdimdim commented:
It's possible. I didn't looked deeply at the F1, my requirements were 
for the F4. When I quickly looked at the F1 datasheet, it seemed close 
enough, but I didn't looked at the details. Still better than nothing ?

Le 05/04/2016 07:47, Liviu Ionescu a écrit :
>
> I took a second look at your USART, it seems very specific to F4, 
> please be aware that other families have different definitions, for 
> example F1 is completely different.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/gnuarmeclipse/qemu/pull/6#issuecomment-205656123>
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

dimdimdimdim commented:
Hi Liviu,

Yes, I had to do that because the status register gets modified _after_ 
the read (RXNE bit). So the read operation returns a value with RXNE 
set, but the RXNE bit is then cleared from the register (at least it 
should be if subsequent reads are made).
The peripheral_register_read_callback function, as you mention, calls 
pre_read, which can be used to adjust values before reading. But then 
the value returned by pre_read is both used to update the internal 
registry value, and as the returned value from the read operation. This 
is not what we want here. The returned value and new registry value are 
actually different. I certainly could have got around this by adding 
some internal state to hold the RXNE bit and mix it in the pre_read, 
then clear the internal RXNE in the pre_read also. But it felt more 
natural to add a post_read, and closer to the reality of what's done in 
the MCU. It also avoids adding additional state fields in the data 
structure. Moreover, the value field of the PeripheralRegisterState 
structure will always hold the correct value, and I imagine it is 
important for the accuracy of the debug info.
But maybe my logic is wrong, I didn't know anything of the internals of 
qemu before doing this, so there may be things that I'm missing.

Regarding the multiple unrelated issued requiring separate pull 
requests, I thought of that when I did it, but it was too late. Sorry. 
I'm new to collaborative development, so I'm learning that as well.

Dimitri


Le 04/04/2016 19:45, Liviu Ionescu a écrit :
>
> Dimitri,
>
> I finally processed your code. Most of it is already in, although not 
> tested and for sure requiring fine tuning.
>
> I had to manually add the patches, since the pull request in it's 
> entirety was not ok.
>
> One major remark: I did not understand the reason why you added the 
> post_read callback, normally the logic should be implemented in the 
> pre_read callback and the value returned. Then the common code should 
> store this value in the register, and the read should return it 
> (masked and shifted) from there. Similarly, but mirrored, for write, 
> the value is written by the user to the register storage and 
> post_write is invoked.
>
> Could you fix these details and post the changes in a separate pull 
> request?
>
> If you find multiple unrelated issues, please post them as distinct 
> pul requests, otherwise one wrong commit will prevent the acceptance 
> of the entire pull request.
>
> Thank you,
>
> Liviu
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/gnuarmeclipse/qemu/pull/6#issuecomment-205412371>
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

ilg-ul commented:
> the status register gets modified _after_ the read (RXNE bit).

aha! I'll investigate and extend the api with the post_read, as you suggested, if this proves to be the solution.

these days I plan to spend some time on qemu, so any other suggestions are welcomed.

ilg-ul commented:
I was thinking more, and realised that the read operation is actually atomic, it includes running the pre_read action and returning the value to the user. what would be the difference of updating the status register in the pre_read callback or in the post_read, as long as the read did not return yet with the value to the user? reading the status register will be done later, in a separate transaction anyway.

however it should be noted that for some peripherals the logic might be more complicated; for example, the status register might be accessed before accessing the data register, and, in this case, the pre_read callback linked to the data register is not yet executed. for the usart device, it looks like the code to get the byte from the host should be executed either in the data or in the status read, and the result be made available to both, but only once. 

btw, although I guess it does not apply here, there is a mechanism to define automatic status bits, that follow other enable bits.

dimdimdimdim commented:
Well, the problem is, that pre_read is called from peripheral_register_read_callback and, if you look at this function, it does:
```
if (state->pre_read) {
       peripheral_register_t new_value;
       new_value = state->pre_read(reg, periph, addr, offset, size);
       state->value &= (new_value & state->readable_bits);
       state->value |= (new_value & state->readable_bits);
}
```
So changing the value fields _within_ pre_read has no effect, because it will be overridden by the caller anyway (see the two lines `state->value &=` / `|=`, which I don't understand how different it is from a simple `state->value = (new_value & state->readable_bits)` by the way...)

Then peripheral_register_read_callback just returns `state->value & state->readable_bits` (after appropriate shortening). So you can't have a return value different from the updated register value. A post_read (or another mechanism, but that does not exist currently) is required to get the behavior of the RXNE bit.

ilg-ul commented:
I added the post_read callback, and simplified the read value assignment, as suggested.

functions specific to F4 were renamed, functions that apply to all families were extended with switch statements.

the code should be more or less similar to what you submitted. could you test it and fix possible bugs I introduced while manually applying your changes?

dimdimdimdim commented:
OK, but i'm not sure I'll be able to do it in a timely manner. I'll try 
to do it this WE, but I can't guarantee.

Le 05/04/2016 23:45, Liviu Ionescu a écrit :
>
> I added the post_read callback, and simplified the read value 
> assignment, as suggested.
>
> functions specific to F4 were renamed, functions that apply to all 
> families were extended with switch statements.
>
> the code should be more or less similar to what you submitted. could 
> you test it and fix possible bugs I introduced while manually applying 
> your changes?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/gnuarmeclipse/qemu/pull/6#issuecomment-205996935>
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

ilg-ul commented:
I already restructured the entire Cortex-M part of the project, by bringing together all files in a `cortexm` folder, and there are other changes planned (like splitting NVIC and removing dependency to GIC), so probably it would be better to wait for the code to be stable before fixing the serial emulation.
